ecs-init/volumes/ecs_volume_plugin.go

```diff
@@ -201,32 +188,148 @@ func deleteMountPath(path string) error {
+		createReq := &driver.CreateRequest{Name: r.Name, Path: vol.Path, Options: vol.Options}
+		if err := volDriver.Create(createReq); err != nil {
+			seelog.Errorf("Volume %s creation failure: %v", r.Name, err)
+			return nil, fmt.Errorf("failed to mount volume %s: %w", r.Name, err)
+		}
+		seelog.Infof("Volume %s mounted successfully", r.Name)
+	}
+
+	// Update state
+	seelog.Infof("Adding mount %s to volume %s", r.ID, r.Name)
+	vol.AddMount(r.ID)
+	if err := a.state.recordVolume(r.Name, vol); err != nil {
+		// State update failed, so roll back the changes made so far to make state consistent
+		seelog.Errorf("Failed to save volume %s, rolling back changes: %v", r.Name, err)
+		vol.RemoveMount(r.ID)
+		if len(vol.Mounts) == 0 {
+			seelog.Warnf("Rolling back mounting of volume %s", r.Name)
+			if err := volDriver.Remove(&driver.RemoveRequest{Name: r.Name}); err != nil {
+				seelog.Errorf("Volume %s removal failure: %v", r.Name, err)
+			}
+		}
+		a.state.recordVolume(r.Name, vol)
+		return nil, fmt.Errorf("mount failed due to an error while saving state: %w", err)
+	}
+
+	// All good
 	return &volume.MountResponse{Mountpoint: vol.Path}, nil
 }
 
 // Unmount implements Docker volume plugin's Unmount Method
 func (a *AmazonECSVolumePlugin) Unmount(r *volume.UnmountRequest) error {
-	a.lock.RLock()
-	defer a.lock.RUnlock()
-	_, ok := a.volumes[r.Name]
+	seelog.Infof("Received unmount request %+v", r)
+
+	// Validate the request
+	if len(r.Name) == 0 {
+		return fmt.Errorf("no volume in the request")
+	}
+	if len(r.ID) == 0 {
+		return fmt.Errorf("no mount ID in the request")
+	}
+
+	// Acquire write lock
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// Find  the volume
+	vol, ok := a.volumes[r.Name]
 	if !ok {
 		seelog.Errorf("Volume %s to unmount is not found", r.Name)
 		return fmt.Errorf("volume %s not found", r.Name)
 	}
+
+	// Get the corresponding volume driver
+	volDriver, err := a.getVolumeDriver(vol.Type)
+	if err != nil {
+		seelog.Errorf("Volume %s removal failure: %v", r.Name, err)
+		return fmt.Errorf("volume %v of type %s is unsupported: %w", r.Name, vol.Type, err)
+	}
+	if volDriver == nil {
+		// this case should not happen normally
+		return fmt.Errorf("no corresponding volume driver found for type %s", vol.Type)
+	}
+
+	// Remove the mount from the volume
+	seelog.Infof("Removing mount %s from volume %s", r.ID, r.Name)
+	if exists := vol.RemoveMount(r.ID); !exists {
+		seelog.Warnf("Mount %s was not found on volume %s, this is a no-op", r.ID, r.Name)
+		return nil
+	}
+
+	// If there are no more mounts left on the volume then unmount the volume from the host
+	if len(vol.Mounts) == 0 {
+		seelog.Infof("No active mounts left on volume %s, unmounting it", r.Name)
+		if err := volDriver.Remove(&driver.RemoveRequest{Name: r.Name}); err != nil {
+			vol.AddMount(r.ID)
+			seelog.Errorf("Failed to unmount volume %v: %v", r.Name, err)
+			return fmt.Errorf("failed to unmount volume %v: %w", r.Name, err)
+		}
```

Comment:
While it makes sense to roll back partial creation, it feels less common to roll back partial deletion. 

For creation, caller might give up on failure, which makes more sense for it to be an "all or none" operation as we do not want orphaned resources. Deletion, on the other hand, is generally best effort and okay with partial work. As long as we make sure the unmount and removal are idempotent requests, I don't see much benefit recreating the resources as a rollback

Reply:
I agree with you and will make this change. Thanks!

Reply:
Made this change.

Reply:
Thanks! Should we remove the `vol.AddMount(r.ID)` (line 300) as well?

Reply:
Yes I missed that. Thank you! Removed.